python/cudf/cudf/_lib/ast.pyx

```diff
@@ -0,0 +1,348 @@
+    ast.Lt: ASTOperator.LESS,
+    ast.Gt: ASTOperator.GREATER,
+    ast.LtE: ASTOperator.LESS_EQUAL,
+    ast.GtE: ASTOperator.GREATER_EQUAL,
+    ast.BitXor: ASTOperator.BITWISE_XOR,
+    # TODO: The mapping of logical/bitwise operators here is inconsistent with
+    # pandas. In pandas, Both `BitAnd` and `And` map to
+    # `ASTOperator.LOGICAL_AND` for booleans, while they map to
+    # `ASTOperator.BITWISE_AND` for integers. However, there is no good way to
+    # encode this at present because expressions can be arbitrarily nested so
+    # we won't know the dtype of the input without inserting a much more
+    # complex traversal of the expression tree to determine the output types at
+    # each node. For now, we'll rely on users to use the appropriate operator.
+    ast.BitAnd: ASTOperator.BITWISE_AND,
+    ast.BitOr: ASTOperator.BITWISE_OR,
+    ast.And: ASTOperator.LOGICAL_AND,
+    ast.Or: ASTOperator.LOGICAL_OR,
+
+    # Unary operators
+    ast.Invert: ASTOperator.BIT_INVERT,
+    ast.Not: ASTOperator.NOT,
+    # TODO: Missing USub, possibility other unary ops?
+}
+
+
+# Mapping between Python function names encode in an ast.Call node and the
+# corresponding libcudf C++ AST operators.
+python_cudf_function_map = {
+    # TODO: Operators listed on
+    # https://pandas.pydata.org/pandas-docs/stable/user_guide/enhancingperf.html#expression-evaluation-via-eval
+    # that we don't support yet:
+    # expm1, log1p, arctan2 and log10.
+    'sin': ASTOperator.SIN,
+    'cos': ASTOperator.COS,
+    'tan': ASTOperator.TAN,
+    'arcsin': ASTOperator.ARCSIN,
+    'arccos': ASTOperator.ARCCOS,
+    'arctan': ASTOperator.ARCTAN,
+    'sinh': ASTOperator.SINH,
+    'cosh': ASTOperator.COSH,
+    'tanh': ASTOperator.TANH,
+    'arcsinh': ASTOperator.ARCSINH,
+    'arccosh': ASTOperator.ARCCOSH,
+    'arctanh': ASTOperator.ARCTANH,
+    'exp': ASTOperator.EXP,
+    'log': ASTOperator.LOG,
+    'sqrt': ASTOperator.SQRT,
+    'abs': ASTOperator.ABS,
+
+    # TODO: Operators supported by libcudf with no Python function analog.
+    # ast.rint: ASTOperator.RINT,
+    # ast.cbrt: ASTOperator.CBRT,
+    # ast.ceil: ASTOperator.CEIL,
+    # ast.floor: ASTOperator.FLOOR,
+}
+
+
+class libcudfASTVisitor(ast.NodeVisitor):
+    """A NodeVisitor specialized for constructing a libcudf expression tree.
+
+    This visitor is designed to handle AST nodes that have libcudf equivalents.
+    It constructs column references from names and literals from constants,
+    then builds up operations. The final result can be accessed using the
+    `expression` property. The visitor must be kept in scope for as long as the
+    expression is needed because all of the underlying libcudf expressions will
+    be destroyed when the libcudfASTVisitor is.
+
+    Parameters
+    ----------
+    col_names : Tuple[str]
+        The column names used to map the names in an expression.
+    """
+    def __init__(self, col_names: Tuple[str]):
+        self.stack = []
```

Comment:
Would a `deque` perform any better here since all we're doing is pushing and popping?

Reply:
I can try it, but I doubt it. Deques are designed to combine the operations of stacks and queues, i.e. fast insertion/removal from both the beginning and the end. In this case, we are only ever pushing to/popping from the end, so the performance concerns with using lists should be moot (insertion/deletion from the beginning of a list is an O(N) op because it requires shifting all the data to maintain contiguity).

Reply:
Tested it. There is no meaningful performance difference between the two.